netplugin/netd.go

```diff
@@ -14,248 +14,295 @@ limitations under the License.
-	flagSet.BoolVar(&opts.version,
-		"version",
-		false,
-		"Show version")
-	flagSet.StringVar(&opts.dbURL,
-		"cluster-store",
-		"etcd://127.0.0.1:2379",
-		"state store url")
-	flagSet.StringVar(&opts.nwDriver,
-		"net-driver",
-		"ovs",
-		"network driver component ovs|vpp (vpp is not supported yet)")
-	flagSet.IntVar(&opts.vxlanUDPPort,
-		"vxlan-port",
-		4789,
-		"VxLAN UDP port number")
+	// post initialization processing
+	ag.PostInit()
 
-	err = flagSet.Parse(os.Args[1:])
-	if err != nil {
-		log.Fatalf("Failed to parse command. Error: %s", err)
+	// handle events
+	if err := ag.HandleEvents(); err != nil {
+		logrus.Infof("Netplugin exiting due to error: %v", err)
+		os.Exit(1)
 	}
+}
 
-	if opts.version {
-		fmt.Printf(version.String())
-		os.Exit(0)
-	}
+func validateConfigs(ctx *cli.Context) (plugin.Config, error) {
 
-	// Make sure we are running as root
-	usr, err := user.Current()
-	if (err != nil) || (usr.Username != "root") {
-		log.Fatalf("This process can only be run as root")
+	// 1. validate and set up log
+	if ctx.Bool("use-syslog") {
+		syslogURL := ctx.String("syslog-url")
+		configureSyslog(syslogURL)
+		logrus.Infof("Set netplugin use syslog config: %v", syslogURL)
+	} else {
+		logrus.Infof("Set netplugin use syslog config: false")
 	}
 
-	if opts.debug {
-		log.SetLevel(log.DebugLevel)
-		os.Setenv("CONTIV_TRACE", "1")
+	logLevel, err := logrus.ParseLevel(ctx.String("log-level"))
+	if err != nil {
+		return plugin.Config{}, err
 	}
+	logrus.SetLevel(logLevel)
+	logrus.Infof("Use netplugin log level: %v", logLevel)
 
-	if opts.jsonLog {
-		log.SetFormatter(&log.JSONFormatter{})
+	if ctx.Bool("use-json-log") {
+		logrus.SetFormatter(&logrus.JSONFormatter{})
+		logrus.Infof("Use netplugin log format: json")
 	} else {
-		log.SetFormatter(&log.TextFormatter{FullTimestamp: true, TimestampFormat: time.StampNano})
+		logrus.SetFormatter(&logrus.TextFormatter{FullTimestamp: true, TimestampFormat: time.StampNano})
+		logrus.Infof("Use netplugin log format: text")
 	}
 
-	if opts.syslog != "" {
-		configureSyslog(opts.syslog)
+	// 2. validate and set plugin mode
+	pluginMode := strings.ToLower(ctx.String("mode"))
+	switch pluginMode {
+	case master.Docker, master.Kubernetes, master.SwarmMode, master.Test:
+		logrus.Infof("Use netplugin mode: %v", pluginMode)
+	case "":
+		return plugin.Config{}, fmt.Errorf("netplugin mode is not set")
+	default:
+		return plugin.Config{}, fmt.Errorf("Unknown netplugin mode: %v", pluginMode)
 	}
 
-	if flagSet.NFlag() < 1 {
-		log.Infof("host-label not specified, using default (%s)", opts.hostLabel)
+	// 3. validate and set network mode
+	networkMode := strings.ToLower(ctx.String("netmode"))
+	switch networkMode {
+	case "vlan", "vxlan":
+		logrus.Infof("Use netplugin network mode: %v", networkMode)
+	case "":
+		return plugin.Config{}, fmt.Errorf("netplugin network mode is not set")
+	default:
+		return plugin.Config{}, fmt.Errorf("Unknown netplugin network mode: %v", networkMode)
 	}
```

Comment:
seems fatal, exit?

Reply:
exit will happen in main func where also provides error info and code

Reply:
all good